### ISSUE TITLE
TPU training runs extremely slowly. 1 step takes 10x amount of time compared to single GPU

### ISSUE THREAD

**Issue by NakedKoala** [PyTorchLightning/pytorch-lightning#7368]
## 🐛 Bug

I ported my pytorch code to pytorch-lightning hoping to use TPU for speed up.  The ported pytorch-lightning  averages 2-3 steps per second on a single GPU. But when I run the exact same code on Colab TPU runtime and change  `pl.Trainer(gpus=1) ` to `pl.Trainer(tpu_cores=8)`, a single step takes 10 - 20 seconds. 

It is hard to make a simple notebook for reproducing the problem as the entire training script is tightly couple with an complex repository.  Also the steps to properly configure conda environment for training script to run is very complex.

Best I can do is to sharing the  pytorch-lightning code in a gist.

https://gist.github.com/NakedKoala/1a2fb843c08d493a3943d87e2b7010f0

### Expected behavior

TPU training is not slower than single GPU training

**Comment by Borda** [comment]
I think we have fixed it on master, mind check latest `pip install pytorch-lightning==1.3.0rc2`
cc: @kaushikb11

**Comment by NakedKoala** [comment]
I tried `pip install pytorch-lightning==1.3.0rc2`.  Did not help. Still taking half-minute per iteration.  Any earlier version ( before this bug was introduced ) that you can suggest ?

**Comment by tchaton** [comment]
Hey @NakedKoala,

I looked at your gist. You are using only 1 tpu cores. Did you try 8 ?

Best,
T.C

**Comment by NakedKoala** [comment]
> Hey @NakedKoala,
> 
> I looked at your gist. You are using only 1 tpu cores. Did you try 8 ?
> 
> Best,
> T.C

Yes. I started with 8 and saw the slowness and tried to debug by using 1 tpu (may be 1 tpu might avoid issues relating to spawning large number of processes )  But it did not speed up.    Even just 1 tpu should not be 10x slower than a single GPU

**Comment by tchaton** [comment]
Yes, that's definitely annoying.

I didn't notice special slowness when trying Lightning Transformers on TPU. I think it was 3-5 times faster for me compared to 2 gpus training.

Mind putting together a reproducible script ?

Best,
T.C

**Comment by NakedKoala** [comment]
> Yes, that's definitely annoying.
> 
> I didn't notice special slowness when trying Lightning Transformers on TPU. I think it was 3-5 times faster for me compared to 2 gpus training.
> 
> Mind putting together a reproducible script ?
> 
> Best,
> T.C

I will put together a public Colab within half a hour

**Comment by NakedKoala** [comment]
> > Yes, that's definitely annoying.
> > I didn't notice special slowness when trying Lightning Transformers on TPU. I think it was 3-5 times faster for me compared to 2 gpus training.
> > Mind putting together a reproducible script ?
> > Best,
> > T.C
> 
> I will put together a public Colab within half a hour

Actually, do you mind sharing you gmail address. I will share my Private colab notebook with you. ( Repro notebook needs to fetch kaggle competition data.  Don't know how to do it without putting my Kaggle credential explicitly in the notebook )

**Comment by NakedKoala** [comment]
> Yes, that's definitely annoying.
> 
> I didn't notice special slowness when trying Lightning Transformers on TPU. I think it was 3-5 times faster for me compared to 2 gpus training.
> 
> Mind putting together a reproducible script ?
> 
> Best,
> T.C

Working on repro notebook right now.
Q:  Do you have a Kaggle account ?  If so, I would just comment out the credential part and let you insert your own.

**Comment by NakedKoala** [comment]
> Yes, that's definitely annoying.
> 
> I didn't notice special slowness when trying Lightning Transformers on TPU. I think it was 3-5 times faster for me compared to 2 gpus training.
> 
> Mind putting together a reproducible script ?
> 
> Best,
> T.C


Here is the reproduction notebook. Please follow the instruction contain within to reproduce. 
Please save the files in the depdendency_files.zip into your colab runtime and copy them to the respective directory ( instruction is in the notebook).

Please let me know if you have trouble running the notebook

Colab:

https://colab.research.google.com/drive/1nVZ6OzyJRyTMouWJiTTyL3KbLcMvm0DY?usp=sharing


Dependency Files:

https://drive.google.com/file/d/1OhlaIRCEmxLMcfFrta99CVN7qL8-wkZs/view?usp=sharing

**Comment by NakedKoala** [comment]
I used `TPUSpawnPlugin` in the Trainer. Here is the printout.  I am unable to decipher this.  Any info regarding the slowness can be recovered from below ? 
```
pt-xla-profiler: TransferFromServerTime too frequent: 449 counts during 6 steps
pt-xla-profiler: Op(s) not lowered: aten::im2col, aten::im2col_backward,  Please open a GitHub issue with the above op lowering requests.

```

**Comment by NakedKoala** [comment]
I also used import `torch_xla.debug.metrics`  to print metric after each training step. 
Here is what gets printed out for 1 step

```
Metric: ExecuteTime
  TotalSamples: 210
  Accumulator: 03s913ms699.706us
  ValueRate: 010ms881.008us / second
  Rate: 0.712402 / second
  Percentiles: 1%=002ms114.601us; 5%=002ms277.371us; 10%=002ms464.266us; 20%=003ms687.511us; 50%=003ms131.265us; 80%=004ms029.872us; 90%=036ms142.378us; 95%=090ms578.779us; 99%=149ms292.477us
Metric: InboundData
  TotalSamples: 488
  Accumulator: 164.12MB
  ValueRate: 569.95KB / second
  Rate: 1.65502 / second
  Percentiles: 1%=0.00B; 5%=4.00B; 10%=96.00B; 20%=96.00B; 50%=1.50KB; 80%=576.00KB; 90%=1.15MB; 95%=2.25MB; 99%=2.25MB
Metric: InputOutputAliasCount
  TotalSamples: 17
  Accumulator: 12547.00
  ValueRate: 55.10 / second
  Rate: 0.0746605 / second
  Percentiles: 1%=1.00; 5%=1.00; 10%=1.00; 20%=1.00; 50%=1.00; 80%=1729.00; 90%=1729.00; 95%=1729.00; 99%=1729.00
Metric: IrValueTensorToXlaData
  TotalSamples: 464
  Accumulator: 03s319ms807.736us
  ValueRate: 011ms688.789us / second
  Rate: 1.49439 / second
  Percentiles: 1%=001ms383.500us; 5%=002ms592.581us; 10%=002ms682.237us; 20%=002ms869.295us; 50%=002ms161.842us; 80%=008ms008.658us; 90%=018ms629.230us; 95%=027ms902.053us; 99%=038ms516.177us
Metric: OutboundData
  TotalSamples: 505
  Accumulator: 335.15MB
  ValueRate: 1.08MB / second
  Rate: 1.62639 / second
  Percentiles: 1%=4.00B; 5%=4.00B; 10%=96.00B; 20%=96.00B; 50%=1.50KB; 80%=576.00KB; 90%=1.15MB; 95%=2.25MB; 99%=9.22MB
Metric: ReleaseDataHandlesTi
  Percentiles: 1%=001ms415.414us; 5%=002ms591.243us; 10%=002ms689.553us; 20%=002ms873.456us; 50%=002ms178.320us; 80%=008ms053.959us; 90%=021ms019.411us; 95%=027ms311.874us; 99%=194ms038.059us
Metric: TransferToServerTransformTime
  TotalSamples: 505
  Accumulator: 347ms810.469us
  ValueRate: 001ms116.931us / second
  Rate: 1.62639 / second
  Percentiles: 1%=103.857us; 5%=126.661us; 10%=137.395us; 20%=147.275us; 50%=210.807us; 80%=436.981us; 90%=919.946us; 95%=001ms075.260us; 99%=007ms229.038us
Counter: CachedCompile
  Value: 169
Counter: CreateCompileHandles
  Value: 41
Counter: CreateDataHandles
  Value: 13287
Counter: CreateXlaTensor
  Value: 59853
Counter: DestroyDataHandles
  Value: 11494
Counter: DestroyXlaTensor
  Value: 57239
Counter: DeviceDataCacheMiss
  Value: 27
Counter: MarkStep
  Value: 12
Counter: ReleaseDataHandles
  Value: 11494
Counter: UncachedCompile
  Value: 41
Counter: XRTAllocateFromTensor_Empty
  Value: 49
Counter: XrtCompile_Empty
  Value: 144
Counter: XrtExecuteChained_Empty
  Value: 144
Counter: XrtExec
  Value: 970
Counter: xla::fill_
  Value: 7
Counter: xla::gelu
  Value: 240
Counter: xla::gelu_backward
  Value: 168
Counter: xla::index_select
  Value: 10
Counter: xla::masked_fill_
  Value: 51
Counter: xla::mean
  Value: 6
Counter: xla::mm
  Value: 3936
Counter: xla::mul
  Value: 2759
Counter: xla::mul_
  Value: 6078
Counter: xla::native_batch_norm
  Value: 730
Counter: xla::native_batch_norm_backward
  Value: 511
Counter: xla::nll_loss_backward
  Value: 7
Counter: xla::nll_loss_forward
  Value: 8
Counter: xla::permute
  Value: 867
Counter: xla::relu
  Value: 30
Counter: xla::select
  Value: 1626
Counter: xla::slice
  Value: 3008
Coun
  Mean: 11s552ms531.926us
  StdDev: 20s626ms451.587us
  Rate: 0.280191 / second
  Percentiles: 25%=137.535us; 50%=044ms303.427us; 80%=18s625ms789.761us; 90%=37s382ms871.872us; 95%=01m03s260ms071.327us; 99%=01m07s957ms502.371us
Metric: XrtExecute
  TotalSamples: 3270
  Accumulator: 47s048ms508.525us
  Mean: 019ms094.347us
  StdDev: 055ms170.870us
  Rate: 3.80403 / second
  Percentiles: 25%=001ms375.058us; 50%=002ms682.463us; 80%=002ms444.386us; 90%=076ms055.945us; 95%=128ms289.253us; 99%=251ms442.059us
Metric: XrtExecutorEvict
  TotalSamples: 0
  Accumulator: nanB
  Mean: nanB
  StdDev: nanB
  Percentiles: 
Metric: XrtReadLiteral
  TotalSamples: 7479
  Accumulator: 07s840ms184.193us
  Mean: 978.244us
  StdDev: 515.489us
  Rate: 3.87585 / second
  Percentiles: 25%=613.653us; 50%=812.726us; 80%=001ms296.944us; 90%=002ms883.298us; 95%=002ms024.700us; 99%=002ms425.886us
Metric: XrtReleaseAllocation
  TotalSamples: 4807
  Accumulator: 01s379ms343.721us
  Mean: 724.887us
  StdDev: 814.953us
  Rate: 4.02388 / second
  Percentiles: 25%=045.239us; 50%=370.298us; 80%=001ms385.083us; 90%=002ms110.651us; 95%=002ms454.791us; 99%=003ms918.296us


  TotalSamples: 505
  Accumulator: 16s433ms739.020us
  ValueRate: 054ms082.532us / second
  Rate: 1.66203 / second
  Percentiles: 1%=001ms154.548us; 5%=001ms263.647us; 10%=001ms352.982us; 20%=002ms520.668us; 50%=002ms939.772us; 80%=008ms871.668us; 90%=017ms787.944us; 95%=027ms694.975us; 99%=248ms443.830us
Metric: TransferToServerTransformTime
  TotalSamples: 505
  Accumulator: 326ms676.678us
  ValueRate: 001ms071.831us / second
  Rate: 1.662 / second
  Percentiles: 1%=075.977us; 5%=117.546us; 10%=127.075us; 20%=136.829us; 50%=184.361us; 80%=389.226us; 90%=926.046us; 95%=001ms102.299us; 99%=007ms001.112us
Counter: CachedCompile
  Value: 169
Counter: CreateCompileHandles
  Value: 41
Counter: CreateDataHandles
  Value: 13287
Counter: CreateXlaTensor
  Value: 59567
Counter: DestroyDataHandles
  Value: 11494
Counter: DestroyXlaTensor
  Value: 56953
Counter: DeviceDataCacheMiss
  Value: 27
Counter: MarkStep
  Value: 12
Counter: ReleaseDataHandles
  Value: 11494
Counter: UncachedCompile
  Value: 41
Counter: XRTAllocateFromTensor_Empty
  Value: 49
Counter: XrtCompil
  Accumulator: 02h48m51s024ms879.837us
  Mean: 11s643ms131.381us
  StdDev: 20s740ms733.091us
  Rate: 0.280356 / second
  Percentiles: 25%=138.210us; 50%=045ms777.329us; 80%=18s788ms563.419us; 90%=37s404ms787.324us; 95%=01m03s411ms989.704us; 99%=01m07s957ms502.371us
Metric: XrtExecute
  TotalSamples: 3272
  Accumulator: 47s181ms084.691us
  Mean: 019ms221.515us
  StdDev: 055ms277.790us
  Rate: 3.77203 / second
  Percentiles: 25%=001ms375.058us; 50%=002ms682.665us; 80%=002ms459.066us; 90%=076ms060.357us; 95%=128ms295.962us; 99%=251ms442.059us
Metric: XrtExecutorEvict
  TotalSamples: 0
  Accumulator: nanB
  Mean: nanB
  StdDev: nanB
  Percentiles: 
Metric: XrtReadLiteral
  TotalSamples: 7480
  Accumulator: 07s842ms923.311us
  Mean: 979.176us
  StdDev: 516.001us
  Rate: 3.84252 / second
  Percentiles: 25%=613.653us; 50%=813.830us; 80%=001ms304.339us; 90%=002ms883.298us; 95%=002ms024.700us; 99%=002ms425.886us
Metric: XrtReleaseAllocation
  TotalSamples: 4826
  Accumulator: 01s404ms646.122us
  Mean: 740.004us
  StdDev: 824.547us
  Rate: 3.98847 / second
  Percentiles: 25%=044.948us; 50%=388.267us; 80%=001ms433.513us; 90%=002ms129.493us; 95%=002ms467.720us; 99%=003ms918.296us


```

**Comment by kaushikb11** [comment]
This issue has been addressed on Lightning Slack and upon profiling, it was discovered that some operations were not XLA supported.

```
pt-xla-profiler: TransferFromServerTime too frequent: 449 counts during 6 steps
pt-xla-profiler: Op(s) not lowered: aten::im2col, aten::im2col_backward,  Please open a GitHub issue with the above op lowering requests.
```

Related issue has been created in [XLA](https://github.com/pytorch/xla/issues/2932).
